### PR TITLE
[docker] Detect CPUs in container correctly

### LINKED ISSUE #8958: Kubernetes: Resource Detection is not correct
### What is the problem?

Ray 0.8.0, 0.8.5

### Reproduction (REQUIRED)
This is simple, use a kubernetes install wherein the pod CPU does not match the host CPU. In the pod, run a python shell (`kubectl exec -it <pod name> -- python`) and try:
`
ray.available_resources()
ray.cluster_resources()
`
These will report the host resources and not the pod resources. The incorrect details are for CPU, memory and GPU.

### LINKED ISSUE #8983: [Ray] Correctly detect CPUs and RAMs in Docker
### Describe your feature request

When running tasks in Docker, Ray detects CPUs and RAM of the host machine, regardless of CPUs and RAM assigned to docker tasks, could you please use `nproc` to detect CPUs and RAM correctly? The detecting method can incur bugs like this

```
The actor or task with ID ffffffffffffffff45b95b1c0100 is pending and cannot currently be scheduled. It requires {CPU: 1.000000} for execution and {CPU: 1.000000} for placement, but this node only has remaining {node:9.146.140.151: 1.000000}, {CPU: 180.000000}, {memory: 198.681641 GiB}, {object_store_memory: 61.474609 GiB}. In total there are 0 pending tasks and 4 pending actors on this node. This is likely due to all cluster resources being claimed by actors. To resolve the issue, consider creating fewer actors or increase the resources available to this Ray cluster. You can ignore this message if this Ray cluster is expected to auto-scale
```

References: 
* https://github.com/ray-project/ray/issues/8326
* https://stackoverflow.com/questions/49151296/how-many-cpus-does-my-docker-container-have/49153003#49153003

### PR DESCRIPTION
**Pull request by ijrsvt**
## Why are these changes needed?
We currently use a special case to detect memory in docker, this uses a special case for CPUs. 

## Related issue number

Closes #8983 
Closes #8958 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Review by rkooo567 — APPROVED** [review]
Nice! Is it fixing the docker resources detection issues?

**Comment by ijrsvt** [comment]
@rkooo567  Yep it is!

**Review by wuisawesome — APPROVED** [review]

**Comment by rkooo567** [comment]
@ijrsvt This seems to break the whole serve tests https://travis-ci.com/github/ray-project/ray/jobs/380426501

**Comment by wuisawesome** [comment]
> @ijrsvt This seems to break the whole serve tests https://travis-ci.com/github/ray-project/ray/jobs/380426501

I think this is a shortcoming of `serve.init`. It's implicitely calling `ray.init` but doesn't allow args like `num_cpus` to be passed to `ray.init`

cc @simon-mo

**Comment by simon-mo** [comment]
Oh you can do 
```
ray.init(args...)
serve.init()
```

We will just use the existing ray cluster.

**Comment by wuisawesome** [comment]
so do we need to update every test/example to do that (should we do that in this PR?)

**Comment by ijrsvt** [comment]
I'm a bit confused why switching to `nproc` breaks these tests :/ @simon-mo

**Comment by wuisawesome** [comment]
I think travis has some containerization behavior. Before, serve was starting ray with all the cpu's on the machine, whereas now each test case is only recognizing one or 2 vcpu's for the container running the test.

**Review by wuisawesome — APPROVED** [review]
@simon-mo you're probably the best person to fix the merge conflicts, then i think you're safe to merge once it's ready

**Comment by simon-mo** [comment]
Yeah I will try to get this in today

**Review by wuisawesome — APPROVED** [review]

**Comment by simon-mo** [comment]
It's failing Tune tests now. @krfricke @richardliaw can you take a look? You might need to explicitly init ray with more cpus to fix it.

**Comment by richardliaw** [comment]
I suspect this is actually going to break all of RLlib too..

**Comment by wuisawesome** [comment]
Let's cross that bridge when we get there. But yeah, we may need to make special note of this bug fix in the release notes or something since it's so easy to overfit test cases to this bug.

**Comment by richardliaw** [comment]
ostensibly this breaks most libraries if not all libraries running Ray on
travis/other CI frameworks. In which case I don’t think it’s as straight
forward as merging this and putting something in the release notes. All of
the api changes so far have not required a massive combthrough of the
tune/rllib tests.

tldr this is a pretty massive breaking change and there should be a softer
migration strategy

**Comment by wuisawesome** [comment]
> ostensibly this breaks most libraries if not all libraries running Ray on travis/other CI frameworks

I think this only breaks travis (and maybe Jenkins?) IFF `ray.init` isn't called properly. I poked around rllib and it looks like it's properly using fixtures to start ray, so it shouldn't break (we also didn't need to change any core tests). I don't think this should affect Azure pipelines (or GHA). 

This is a breaking change only when you rely on buggy behavior, and if this autodetecting resources in docker is a common use case, I think that just strengthens the argument for doing this before Ray 1.0.

If you feel strongly about this, then maybe we can add a warning log if we detect that we're in docker? 

There are many 1.0 changes that aren't even API compatible with 0.8.7 so I don't think we should try to be bug-compatible.

**Comment by richardliaw** [comment]
My point isn't to be bug-compatible (which makes no sense, yes I agree) or to block this PR - my point is that this change of resource detection can have complicated impacts. And we see it already in our testing infra. 

@ijrsvt can you please add a warning to be logged *once* if cpu_count does not match the output of nproc and say:

1. what are probable causes
2. what are potential workarounds

File: python/ray/utils.py [hunk 1]
```diff
@@ -489,6 +490,15 @@ def get_system_memory():
     return psutil_memory_in_bytes
 
 
+def get_num_cpus():
+    try:
+        return int(subprocess.check_output("nproc"))
```
Comment (richardliaw) [inline python/ray/utils.py:495]: consider also allowing this behavior to be overriden with an env var, like `RAY_DETECT_NPROC=0`

Comment (rkooo567) [reply python/ray/utils.py:495]: I am on the same page we should probably have more warning msgs in case this change broke things silenty, but it doesn’t make sense to me to have an option to turn it off. Is it for isolating user issues in the future?

**Comment by rkooo567** [comment]
I might have misunderstood what breaks tune / rllib, so bear with me if I am saying something wrong. Imo, the impact of this change is not actually really big. 

- If users don't specify num_cpus => This change will ensure right behaviors.
- If users specified num_cpus => Users do it only when they need fixed number of processes or resources, and it won't break anything when they use Docker.

I think the only case when it can be harmful is when users have tests with unfixed number of cpus, but their test result depends on it, which is very rare case outside our CI.

File: python/ray/utils.py (hunk 1 above)
Comment (richardliaw) [reply python/ray/utils.py:495]: Think about it like this - effectively, `multiprocessing.cpu_count()` is "broken". But by pushing this change, in a vague sense, Ray becomes "incompatible with multiprocessing, psutil, and os.cpu_count". It is also likely that the world revolves around multiprocessing and its behaviors, rather than Ray.
 
Also, think about the person who is developing a library on top of Modin/or some other library. They do not even know that ray even exists. Suddenly their entire codebase breaks - should they pin ray to older version? should they stick "ray.init" everywhere in their code?

File: python/ray/utils.py [hunk 2]
```diff
@@ -489,6 +490,15 @@ def get_system_memory():
     return psutil_memory_in_bytes
 
 
+def get_num_cpus():
+    try:
```
Comment (richardliaw) [inline python/ray/utils.py:494]: also, does https://stackoverflow.com/a/55423170/2213289 sched_affinity work too? would that be more pythonic?

Comment (wuisawesome) [reply python/ray/utils.py:494]: haven't tested, but docker usually requires some wrangling to get affinity to work. (also the `os` module and docker just generally don't play nice).

File: python/ray/utils.py [hunk 3]
```diff
@@ -489,6 +490,31 @@ def get_system_memory():
...
+    cpu_count = multiprocessing.cpu_count()
+    try:
+        # Not easy to get cpu count in docker, see:
+        # https://bugs.python.org/issue36054
+        docker_count = subprocess.check_output("nproc")
+        if docker_count != cpu_count:
+            cpu_count = docker_count
+            if "RAY_DISABLE_DOCKER_CPU_WARNING" not in os.environ:
+                logger.warning(
+                    "Detecting limited number of GPUs due to docker. In "
```
Comment (yiranwang52) [inline python/ray/utils.py:503]: Typo? GPU -> CPU

**Comment by yiranwang52** [comment]
I think instead of Running nproc, we should either use cpu_quota (like how is handled by some libraries in [Go](https://github.com/uber-go/automaxprocs/blob/946a8391268aea0a60a86403988ff3ab4b604a83/internal/runtime/cpu_quota_linux.go#L33)) or cpu_quota/cpu_period (like how this is handled in [JDK](https://bugs.openjdk.java.net/browse/JDK-8146115)).

**Comment by sven1977** [comment]
taking a look at the RLlib test failures. ...

**Comment by sven1977** [comment]
@wuisawesome could you check the failing RLlib, Tune, SGD, and doc tests? They all seem to fail on:
```
File "/home/travis/.cache/bazel/_bazel_travis/b88c129a127452fc94033a29d9f90e20/execroot/com_github_ray_project_ray/bazel-out/k8-opt/bin/rllib/test_env_with_subprocess.runfiles/com_github_ray_project_ray/rllib/tests/test_env_with_subprocess.py", line 31, in <module>
3661    ray.init()
3662  File "/home/travis/build/ray-project/ray/python/ray/worker.py", line 699, in init
3663    ray_params=ray_params)
3664  File "/home/travis/build/ray-project/ray/python/ray/node.py", line 207, in __init__
3665    self.start_ray_processes()
3666  File "/home/travis/build/ray-project/ray/python/ray/node.py", line 813, in start_ray_processes
3667    self.start_raylet()
3668  File "/home/travis/build/ray-project/ray/python/ray/node.py", line 733, in start_raylet
3669    object_spilling_config=self._ray_params.object_spilling_config)
3670  File "/home/travis/build/ray-project/ray/python/ray/services.py", line 1336, in start_raylet
3671    static_resources = resource_spec.to_resource_dict()
3672  File "/home/travis/build/ray-project/ray/python/ray/resource_spec.py", line 110, in to_resource_dict
3673    or isinstance(resource_quantity, float))
3674AssertionError
3675
```

and

```
resources = {
3325            resource_label: resource_quantity
3326            for resource_label, resource_quantity in resources.items()
3327            if resource_quantity != 0
3328        }
3329    
3330        # Check types.
3331        for resource_label, resource_quantity in resources.items():
3332            assert (isinstance(resource_quantity, int)
3333>                   or isinstance(resource_quantity, float))
3334E           AssertionError
3335
3336/home/travis/build/ray-project/ray/python/ray/resource_spec.py:110: AssertionError
```

File: python/ray/utils.py (hunk 1 above)
Comment (wuisawesome) [reply python/ray/utils.py]: Btw, libraries which break, under these conditions also break if they're run on bare metal (or a vm) with 1 cpu. It should probably be a best practice for libraries to check that they have enough CPU's available. Modin, for example, would not have to do this because it works with 1 CPU, whereas many of our libraries would not.

**Comment by wuisawesome** [comment]
@sven1977 @richardliaw I think all the remaining tests failures are about insufficient resources. Can you two fix those tests (and any corresponding documentation that's needed)?

**Comment by richardliaw** [comment]
All tune/sgd tests passing.

**Comment by richardliaw** [comment]
@ijrsvt @sven1977 please open a separate issue to actually properly convert rllib tests. Thanks!
